### ISSUE TITLE
"create missing override" imports from `lib/src` instead of the public library 

### ISSUE THREAD

**Issue by osaxma** [Dart-Code/Dart-Code#4902]
**Describe the bug**

In short:

https://github.com/Dart-Code/Dart-Code/assets/46427323/79964353-e203-422f-8e12-1d7d29788728

**To Reproduce**
```dart
// file1.dart
import 'package:flutter/material.dart';

abstract class X {
  Widget getWidget();
}
```

```dart
// file2.dart
import 'file1.dart';

class Y extends X { // <~~ choose `convert missing overrides` fix  

}
```

**Expected behavior**

I would expect the import to be from a public library (i.e. `package:flutter/widgets.dart` )

-----

<details>

<summary><strong>Workspace Environment</strong></summary>

```text
Dart Code extension: 3.78.2
Flutter extension: 3.78.0 (activated)

App: Visual Studio Code
App Host: desktop
Version: mac 1.85.0

Workspace type: Flutter (LSP)

Dart (3.2.0): ~/fvm/versions/stable/bin/cache/dart-sdk
Flutter (3.16.0): ~/fvm/versions/stable (macOS (darwin/macos))

```

</details>

<details>

<summary><strong>Output from 'dart info'</strong></summary>

`/Users/osaxma/fvm/versions/stable/bin/cache/dart-sdk/bin/dart info`

If providing this information as part of reporting a bug, please review the information
below to ensure it only contains things you're comfortable posting publicly.

#### General info

- Dart 3.2.0 (stable) (Tue Nov 14 18:26:59 2023 +0000) on "macos_x64"
- on macos / Version 12.7 (Build 21G816)
- locale is en

#### Project info

- sdk constraint: '>=3.1.3 <4.0.0'

#### Process info

|  Memory |  CPU | Elapsed time | Command line                                                                               |
| ------: | ---: | -----------: | ------------------------------------------------------------------------------------------ |
|    6 MB | 0.0% |  05-01:31:50 | dart devtools --machine --allow-embedding                                                  |
|  942 MB | 0.0% |  05-01:31:50 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.78.2            |
| 3986 MB | 0.0% |     11:23:38 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.78.2            |
|   15 MB | 0.4% |  05-01:31:50 | flutter_tools.snapshot daemon                                                              |
|   54 MB | 0.0% |  05-00:11:39 | flutter_tools.snapshot debug_adapter                                                       |
|   88 MB | 0.0% |  05-00:11:38 | flutter_tools.snapshot run --machine --start-paused --dart-define DEBUG=true -d macos --devtools-server-address http:<path>/ --target <path>/main.dart |
|  312 MB | 0.0% |  05-00:11:37 | frontend_server.dart.snapshot --sdk-root <path>/ --incremental --target=flutter --experimental-emit-debug-metadata -DDEBUG=true -DFLUTTER_WEB_AUTO_DETECT=true -DFLUTTER_WEB_CANVASKIT_URL=https:<path>/ --output-dill <path>/app.dill --packages <path>/package_config.json -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts --track-widget-creation --filesystem-scheme org-dartlang-root --initialize-from-dill build/3cee5f2e23966d44bcc8c5ed0cf737ea.cache.dill.track.dill --verbosity=error --enable-experiment=alternative-invalidation-strategy |

</details>

<details>

<summary><strong>Output from 'flutter doctor'</strong></summary>

`/Users/osaxma/fvm/versions/stable/bin/flutter doctor -v`

```text
[✓] Flutter (Channel stable, 3.16.0, on macOS 12.7 21G816 darwin-x64, locale en)
    • Flutter version 3.16.0 on channel stable at /Users/osaxma/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db7ef5bf9f (5 weeks ago), 2023-11-15 11:25:44 -0800
    • Engine revision 74d16627b9
    • Dart version 3.2.0
    • DevTools version 2.28.2

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/osaxma/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.85.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.7 21G816 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.109

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by DanTup** [comment]
Sorry for the delay, I somehow missed this.

This seems similar to https://github.com/dart-lang/sdk/issues/49081. The file that contains the definition is being added but we should search for a public import.

We may need to consider the performance of this though, as the edits for this fix are probably computed up-front (and not only once you select the fix).

**Comment by DanTup** [comment]
@bwilkerson slightly related a discussion we had last week.. Today, fixes like "create missing override" are computed up-front and included inline in the `textDocument/codeActions` response. Fixing the issue above will likely mean this fix will search libraries to find the preferred `lib/` import for a type included in the override signatures (which in my testing is fast, but I suspect on very large projects could be less fast).

It would be nice if we could produce the list of available fixes (for `textDocument/codeActions`) without necessarily fully computing the edits (but still accurately deciding that a fix applies) and then "later" compute the fixes. I don't know how complicated that would be (probably it would involves changes in the fix processors that currently just return the computed fixes and not a list of which fixes apply). I also don't know how many places we have "expensive" work like this we might want to try to avoid putting inline into the `/codeActions` request).

Interested in your thoughts/opinion.

**Comment by bwilkerson** [comment]
I think that the first thing we need to do is to understand just how efficient or inefficient such an algorithm would be. To that end, I've opened https://github.com/dart-lang/sdk/issues/55480 as a place for discussion.

We could certainly change `CorrectionProducer` so that it could optionally postpone the creation of the edits, but the legacy protocol doesn't provide a way for those edits to be requested after the fact. I'm not sure whether the IntelliJ APIs would support that mode, even if we extended the legacy protocol to support it (or tried to move to LSP for code actions).
